store/sqlstore/thread_store.go

```diff
@@ -677,29 +667,59 @@ func (s *SqlThreadStore) MarkAllAsReadByTeam(userId, teamId string) error {
 
 // MarkAsRead marks the given thread for the given user as unread from the given timestamp.
 func (s *SqlThreadStore) MarkAsRead(userId, threadId string, timestamp int64) error {
-	query, args, err := s.getQueryBuilder().
-		Update("ThreadMemberships").
-		Where(sq.Eq{"UserId": userId}).
-		Where(sq.Eq{"PostId": threadId}).
-		Set("LastViewed", timestamp).
-		Set("LastUpdated", model.GetMillis()).
+	transaction, err := s.GetMasterX().Beginx()
+	if err != nil {
+		return errors.Wrap(err, "begin_transaction")
+	}
+	defer finalizeTransactionX(transaction)
+
+	queryString, args, err := s.getQueryBuilder().
+		Select("COUNT(Posts.Id)").
+		From("Posts").
+		Where(sq.And{
+			sq.Eq{"Posts.RootId": threadId},
+			sq.Eq{"Posts.DeleteAt": 0},
+			sq.LtOrEq{"Posts.CreateAt": timestamp},
+		}).
```

Comment:
This method does dual duty:

if  the `timestamp` argument is greater than last reply at of the thread, then it marks the thread as read.

If the `timestamp` argument is less that the last reply at of the thread, then it marks the thread as unread after that timestamp.


In the first case, we shouldn't have to scan the posts table, we can simply assign `ThreadMemberships.SeenReplyCount = Threads.ReplyCount`. 
The client can't tell us if its seen the whole thread, because it may have out-of-date information about the thread, i.e. it may not have the latest `LastReplyAt` of the Thread. 

We could check if the `timestamp` is after the `LastReplyAt` of the thread. In this case we set `ThreadMemberships.SeenReplyCount = Threads.ReplyCount`. In the complementary case, we do the posts table scan as above. 

I'm still investigating if this extra check is going to give us a performance win over always scanning the posts table.

Reply:
If we inline things, can we do it all in a `SELECT CASE WHEN ...`?